slangpy/torchintegration/torchtensormarshall.py

```diff
@@ -1 +1,272 @@
+            raise ValueError(
+                f"Tensor shape {shape} does not match expected shape {binding.vector_type.shape}"
+            )
+        assert primal.is_cuda
+
+        if context.device.info.type != DeviceType.cuda:
+
+            data_type = _torch_to_data_type[self.torch_dtype]
+            data.interop_buffer = get_storage(context, primal.numel(), primal.element_size())
+
+            interop_tensor = cast(
+                torch.Tensor,
+                data.interop_buffer.to_torch(type=data_type, shape=shape, strides=strides),
+            )
+            interop_tensor.copy_(primal)
+
+            primal_calldata = {
+                "buffer": data.interop_buffer,
+                "layout": {"offset": 0, "strides": strides},
+                "_shape": shape,
+            }
+
+            if not self.d_in and not self.d_out:
+                return primal_calldata
+
+            result = {"primal": primal_calldata}
+            if self.d_in is not None:
+                if data.grad_in is None:
+                    raise ValueError("Missing required input gradients")
+                result["d_in"] = self.d_in.create_calldata(context, binding, data.grad_in)
+            if self.d_out is not None:
+                if data.grad_out is None:
+                    raise ValueError("Missing tensor to hold output gradients")
+                result["d_out"] = self.d_out.create_calldata(context, binding, data.grad_out)
+
+            if (
+                context.call_mode != CallMode.prim
+                and data.grad_in is not None
+                and data.grad_in is data.grad_out
+            ):
+                if binding.access[1] == AccessType.readwrite:
+                    raise ValueError(
+                        "inout parameter gradients need separate buffers for inputs and outputs (see Tensor.with_grads)"
+                    )
+        else:
+            primal_calldata = {
+                "buffer": primal.data_ptr(),
+                "layout": {"offset": 0, "strides": strides},
+                "_shape": shape,
+            }
+
+            if not self.d_in and not self.d_out:
+                return primal_calldata
+
+            result = {"primal": primal_calldata}
+            if self.d_in is not None:
+                if data.grad_in is None:
+                    raise ValueError("Missing required input gradients")
+                result["d_in"] = self.d_in.create_calldata(context, binding, data.grad_in)
+            if self.d_out is not None:
+                if data.grad_out is None:
+                    raise ValueError("Missing tensor to hold output gradients")
+                result["d_out"] = self.d_out.create_calldata(context, binding, data.grad_out)
+
+        return result
+
+    def read_calldata(
+        self,
+        context: CallContext,
+        binding: "BoundVariableRuntime",
+        data: TensorRef,
+        result: Any,
+    ):
+        if context.device.info.type != DeviceType.cuda:
```

Comment:
This should also check whether `binding.access[0]` is `write` or `readwrite`. Currently it will also copy back all read-only arguments.